MaxText/layers/attentions.py

```diff
@@ -260,22 +241,174 @@ def key_rotary(self, key: Array, inputs_positions: Array):
+            axis_names,
+            segment_axis_names,
+        ),
+        out_specs=axis_names,
+        check_rep=False,
+    )
+    def wrap_flash_attention(query, key, value, decoder_segment_ids):
+      if decoder_segment_ids is not None:
+        assert (
+            query.shape[2]
+            == decoder_segment_ids.q.shape[1]
+        ), 'Sharding along sequence dimension not allowed in flash attention'
+      return tpu_flash_attention.flash_attention(
+          query,
+          key,
+          value,
+          causal=True,
+          segment_ids=decoder_segment_ids,
+          block_sizes=tpu_flash_attention.BlockSizes(
+              block_q=min(512, query.shape[2]),
+              block_k_major=min(512, key.shape[2]),
+              block_k=min(512, key.shape[2]),
+              block_b=min(2, query.shape[0]),
+              block_q_major_dkv=min(512, query.shape[2]),
+              block_k_major_dkv=min(512, key.shape[2]),
+              block_q_dkv=min(512, query.shape[2]),
+              block_k_dkv=min(512, key.shape[2]),
+              block_q_dq=min(1024, query.shape[2]),
+              block_k_dq=min(256, key.shape[2]),
+              block_k_major_dq=min(512, key.shape[2]),
+          ),
+      )
+
+    devices_in_data_fsdp = self.mesh.shape['data'] * self.mesh.shape['fsdp']
+    assert (query.shape[0] / devices_in_data_fsdp).is_integer(), (
+        'Batch dimension should be shardable among the devices in data and fsdp'
+        ' axis'
+    )
+    x = wrap_flash_attention(query, key, value, decoder_segment_ids)
+    x = jax.numpy.transpose(x, axes=(0, 2, 1, 3))
+    return x
+
+  def gpu_flash_attention(
+      self,
+      query: Array,
+      key: Array,
+      value: Array,
   ) -> Array:
-    """Apply Attention."""
+    """GPU Flash Attention."""
+    b, n, s, h = key.shape # pylint: disable=unused-variable
+    if self.attention_kernel == "gpu_flash_xla":
+      bwd_pass_impl = "xla"
+    elif self.attention_kernel == "gpu_flash_triton":
+      bwd_pass_impl = "triton"
+    else:
+      raise ValueError(f"Can't convert {self.attention_kernel } to a bwd_pass_impl")
+  
+    bwd_pass_impl = self.config.gpu_flash_attention_backward_pass_impl
+    axis_names = nn.logical_to_mesh_axes(self.flash_axis_names)
+    segment_axis_names = nn.logical_to_mesh_axes((BATCH, LENGTH))
 
-    del decoder_segment_ids
-    del decode
+    @functools.partial(
+        shard_map,
+        mesh = self.mesh,
+        in_specs = (
+          axis_names,
+          axis_names,
+          axis_names,
+          segment_axis_names),
+        out_specs = axis_names,
+        check_rep=False)
+    def wrap_gpu_flash_attention(query, key, value):
+      return pallas_attention.mha(
+        query, key, value, sm_scale=1.0 / math.sqrt(h), backward_pass_impl=bwd_pass_impl,
+        num_stages = 1, causal = True, segment_ids = None
+      )
+    return wrap_gpu_flash_attention(query, key, value)
 
+  def apply_attention_dot(
```

Comment:
Incomplete docstring.

Reply:
Legitimate feedback but I'm just trying to move things around here! Thanks for calling it out.